assets/tests/TaskSchemaValidationService.js

```diff
@@ -0,0 +1,225 @@
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service for validating data put into a task schema
+ * within a question.
+ */
+
+tie.factory('TaskSchemaValidationService', ['SYSTEM_CODE',
+  function(SYSTEM_CODE) {
+    var AUXILIARY_CODE_CLASS_NAME = 'AuxiliaryCode.';
+    var SYSTEM_CODE_CLASS_NAME = 'System.';
+    // TODO(eyurko): Update this once we support nonlinear runtimes.
+    var ALLOWED_RUNTIMES = ['linear'];
+    var _checkIfFunctionExistsInCode = function(functionName, code) {
+      var functionDefinition = 'def ' + functionName;
+      return code.indexOf(functionDefinition) !== -1;
+    };
+    var _checkIfFunctionExistsInClass = function(functionName, 
+      className, code) {
+      if (functionName.startsWith(className)) {
+        var strippedFunctionName = functionName.substring(
+          className.length);
+        return _checkIfFunctionExistsInCode(
+          strippedFunctionName, code);
+      }
+      return false;
+    };
+    var taskSchemaValidator = {
+      _starterCode: '',
+      _auxiliaryCode: '',
+      getAllFunctions: function() {
+        var functions = [];
+        for (var l in this) {
+          if (this.hasOwnProperty(l) && 
+              this[l] instanceof Function &&
+              !(/functions/i).test(l) &&
+              l.toString().startsWith('verify')) {
+            functions.push(this[l]);
+          }
+        }
+        return functions;
+      },
+      init: function(starterCode, auxiliaryCode) {
+        this._starterCode = starterCode;
+        this._auxiliaryCode = auxiliaryCode;
+      },
+      verifyInstructionsAreNotEmpty: function(task) {
+        var instructions = task.getInstructions();
+        return instructions.length > 0;
+      },
+      verifyInstructionsIsArrayOfStrings: function(task) {
+        var instructions = task.getInstructions();
+        return (angular.isArray(instructions) &&
+          instructions.every(function(instruction) {
+            return angular.isString(instruction);
+          }));
+      },
+      verifyMainFunctionNameIsString: function(task) {
+        var mainFunctionName = task.getMainFunctionName();
+        return (angular.isString(mainFunctionName) &&
+            mainFunctionName.length > 0);
+      },
+      verifyMainFunctionNameAppearsInStarterCode: function(
+        task) {
+        var mainFunctionName = task.getMainFunctionName();
+        return _checkIfFunctionExistsInCode(mainFunctionName, 
+          taskSchemaValidator._starterCode);
+      },
+      verifyInputFunctionNameIsNullOrString: function(task) {
+        var inputFunctionName = task.getInputFunctionName();
+        return (inputFunctionName === null ||
+          angular.isString(inputFunctionName));
```

Comment:
Need len > 0? (for consistency with main func name)

Ditto for output func name below

I can see a rationale for not doing the above though -- and that is if you're thinking, input/output func name is being checked within the code itself. That sounds fine to me. But if you do that then maybe break the len > 0 check for main func name out into its own separate test so that we have parity between the three of them.